src/parser.java

```diff
@@ -0,0 +1,214 @@
+package src;
```

Comment:
I know this is the meat and potatoes of this PR, but it's super hard check if this is right just by reading. I think we should work towards easier testing so that we can confidently verify this.